OpenRA.Mods.Common/Traits/Buildings/FreeActor.cs

```diff
@@ -28,23 +28,38 @@ public class FreeActorInfo : ITraitInfo
 		[Desc("Which direction the unit should face.")]
 		public readonly int Facing = 0;
 
-		public virtual object Create(ActorInitializer init) { return new FreeActor(init, this); }
+		public override object Create(ActorInitializer init) { return new FreeActor(init, this); }
 	}
 
-	public class FreeActor
+	public class FreeActor : ConditionalTrait<FreeActorInfo>
 	{
-		public FreeActor(ActorInitializer init, FreeActorInfo info)
+		FreeActorInfo info;
+
+		public FreeActor(ActorInitializer init, FreeActorInfo info) : base(info)
 		{
 			if (init.Contains<FreeActorInit>() && !init.Get<FreeActorInit>().ActorValue)
 				return;
 
-			init.Self.World.AddFrameEndTask(w =>
+			this.info = info;
+
+			if (!IsTraitDisabled)
+				GrantFreeActor(init.Self);
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			GrantFreeActor(self);
```

Comment:
Didn't test, but i think with current code it would trigger every time trait gets enabled. I guess we would want that to fire only once.

Reply:
Yes and I don't see that as an issue, moreso as a feature. If someone wants this only to be enabled once then said one should never disable the trait. There's no issue here.

Reply:
> Edit: Or add a boolean to set if it will fire once or every time when enabled.

Good idea.  Can we do this?

Reply:
Sure, will do!